### ISSUE TITLE
Add `BuildInfo` provider to rules_rust public API

### ISSUE THREAD

**Issue by UebelAndre** [bazelbuild/rules_rust#1578]
This ticket expands on the thread started here: https://github.com/bazelbuild/rules_rust/pull/1537#issuecomment-1256297189

There have been a couple pull-requests around [cargo_build_script](https://github.com/bazelbuild/rules_rust/blob/0.10.0/cargo/cargo_build_script.bzl#L311-L435) which increase it's footprint the core Rust rules. I don't think the Bazel rules should become too entangled with the concept of [Cargo build scripts](https://doc.rust-lang.org/cargo/reference/build-scripts.html) but I do think it would make sense to allow users to define their own rules and like `cargo_build_script` and [`cargo_dep_env`](https://github.com/bazelbuild/rules_rust/blob/0.10.0/cargo/cargo_build_script.bzl#L474-L498) to wire up regular `rustc` flags. [BuildInfo](https://github.com/bazelbuild/rules_rust/blob/0.10.0/rust/private/providers.bzl#L59-L69) seems to be a sufficient interface for this and I think it'd be better to try and understand what the provider or core rules lack to keep them "Cargo free" (to clarify, Cargo is Rust's primary build system so i'm of the opinion that Bazel should match the behaviors of Cargo within reason without compromising on the premise that the Bazel rules are primarily focused on invoking `rustc` in a similar way as Cargo). 

One could also think of this issue as "How to remove `# buildifier: disable=bzl-visibility` from [cargo_build_script.bzl](https://github.com/bazelbuild/rules_rust/blob/0.10.0/cargo/cargo_build_script.bzl#L7-L14) as I think there'd be a ton more freedom in making the rules defined there work more correctly if improving them didn't require updating any core Rust rules.

Related Issues/Pull-Requests:
- https://github.com/bazelbuild/rules_rust/pull/1551
- https://github.com/bazelbuild/rules_rust/pull/1537

**Comment by UebelAndre** [comment]
cc @illicitonion @scentini @krasimirgg

Do you folks have any thoughts on a path forward here aside from what as mentioned on https://github.com/bazelbuild/rules_rust/pull/1537#issuecomment-1256297189 ?

**Comment by krasimirgg** [comment]
I'm not super familiar with cargo_build_script's internals, but here are my thoughts about the private visibility instances in cargo_build_script.bzl: 
* It feels like `BuildInfo` should be turned into a public API. The `DepInfo` not so much, I wonder if we can enhance the BuildInfo to support the use case where DepInfo is used...
* The `get_compilation_mode_opts` and `get_linker_and_args` seem generally useful enough to be used as a basis for some public API functionality, but currently they are not up to par: for `get_compilation_mode_opts`, we either have a more specific `compilation_mode_opt_level()` public API, or define a proper `compilation_mode_opts` public struct. For `get_linker_and_args`, we'd need to not depend on the attr argument and instead define a proper argument object to hold and pass in the dependencies.
* The ones from `rust/private:utils.bzl` don't feel like they should be made public to me.

**Comment by allvpv** [comment]
yes, specifically we would like to access the `target[BuildInfo].out_dir` property and pass it to 'intellij-rust'

the 'rust_analyzer.bzl' also accesses this field, but this file is a part of this repo, so the provider being private is not a problem there. to achieve the same functionality in our case, we need to have this provider accessible publicly

we cannot use the existing 'gen_rust_project' tool, which generates a `rust_project.json` file, because it is not very well suited for us (the generated JSON is stripped of Bazel target labels, etc) and our objective is to retrieve all necessary data directly from Bazel

**Comment by illicitonion** [comment]
I'd be happy to make it public as-is, all of the concepts on it are pretty generic.

@scentini @krasimirgg @UebelAndre any objections, or API changes you'd like to make before doing so?

**Comment by scentini** [comment]
Given that we have yet to reach 1.0, everything that is public is still unstable :) So I don't have any objections.

**Comment by abrams27** [comment]
thanks a lot!
